File: 3870-minimum-moves-to-clean-the-classroom/minimum-moves-to-clean-the-classroom.py
from collections import deque
from typing import List

class Solution:
    def minMoves(self, classroom: List[str], energy: int) -> int:
        m, n = len(classroom), len(classroom[0])
        start_r, start_c = -1, -1
        litter_locations = []
        
        for r in range(m):
            for c in range(n):
                cell = classroom[r][c]
                if cell == 'S':
                    start_r, start_c = r, c
                elif cell == 'L':
                    litter_locations.append((r, c))
                    
        num_litter = len(litter_locations)
        if num_litter == 0:
            return 0
            
        target_mask = (1 << num_litter) - 1
        litter_map = {pos: i for i, pos in enumerate(litter_locations)}
        
        best_energy = {}
        queue = deque([(start_r, start_c, 0, energy, 0)])
        best_energy[(start_r, start_c, 0)] = energy
        
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        
        while queue:
            r, c, mask, cur_energy, steps = queue.popleft()
            
            if mask == target_mask:
                return steps
                
            if cur_energy == 0:
                continue
                
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                
                if 0 <= nr < m and 0 <= nc < n and classroom[nr][nc] != 'X':
                    next_cell = classroom[nr][nc]
                    next_mask = mask
                    next_energy = cur_energy - 1
                    
                    if next_cell == 'L' and (nr, nc) in litter_map:
                        next_mask |= (1 << litter_map[(nr, nc)])
                        
                    if next_cell == 'R':
                        next_energy = energy
                        
                    state = (nr, nc, next_mask)
                    if next_energy > best_energy.get(state, -1):
                        best_energy[state] = next_energy
                        queue.append((nr, nc, next_mask, next_energy, steps + 1))
                        
        return -1

File: 3870-minimum-moves-to-clean-the-classroom/minimum-moves-to-clean-the-classroom.java
class Solution {
    static class State {
        int r, c, mask, energy, steps;
        State(int r, int c, int mask, int energy, int steps) {
            this.r = r;
            this.c = c;
            this.mask = mask;
            this.energy = energy;
            this.steps = steps;
        }
    }

    public int minMoves(String[] classroom, int energy) {
        int m = classroom.length;
        int n = classroom[0].length();
        int startR = -1, startC = -1;
        List<int[]> litterLocations = new ArrayList<>();

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                char ch = classroom[r].charAt(c);
                if (ch == 'S') {
                    startR = r;
                    startC = c;
                } else if (ch == 'L') {
                    litterLocations.add(new int[]{r, c});
                }
            }
        }

        int numLitter = litterLocations.size();
        if (numLitter == 0) return 0;

        int targetMask = (1 << numLitter) - 1;
        int[][] litterIdxMap = new int[m][n];
        for (int i = 0; i < m; i++) Arrays.fill(litterIdxMap[i], -1);
        for (int i = 0; i < numLitter; i++) {
            int[] loc = litterLocations.get(i);
            litterIdxMap[loc[0]][loc[1]] = i;
        }

        int[][][] bestEnergy = new int[m][n][1 << numLitter];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                Arrays.fill(bestEnergy[i][j], -1);
            }
        }

        Queue<State> queue = new LinkedList<>();
        queue.add(new State(startR, startC, 0, energy, 0));
        bestEnergy[startR][startC][0] = energy;

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!queue.isEmpty()) {
            State curr = queue.poll();

            if (curr.mask == targetMask) return curr.steps;
            if (curr.energy == 0) continue;

            for (int[] dir : dirs) {
                int nr = curr.r + dir[0];
                int nc = curr.c + dir[1];

                if (nr >= 0 && nr < m && nc >= 0 && nc < n && classroom[nr].charAt(nc) != 'X') {
                    char cell = classroom[nr].charAt(nc);
                    int nextMask = curr.mask;
                    int nextEnergy = curr.energy - 1;

                    if (cell == 'L' && litterIdxMap[nr][nc] != -1) {
                        nextMask |= (1 << litterIdxMap[nr][nc]);
                    }

                    if (cell == 'R') {
                        nextEnergy = energy;
                    }

                    if (nextEnergy > bestEnergy[nr][nc][nextMask]) {
                        bestEnergy[nr][nc][nextMask] = nextEnergy;
                        queue.add(new State(nr, nc, nextMask, nextEnergy, curr.steps + 1));
                    }
                }
            }
        }

        return -1;
    }
}